src/mobster/cmd/generate/oci_image/spdx_utils.py

```diff
@@ -0,0 +1,317 @@
+    if supplier := package.get("supplier"):
+        package["supplier"] = await normalize_actor(supplier)
+
+
+async def normalize_sbom(sbom: dict[str, Any]) -> None:
+    """
+    Adds necessary fields to an SPDX SBOM to be loaded by the
+    SPDX library without validation issues.
+    Args:
+        sbom (dict[str, Any]): The SBOM to be normalized.
+
+    Returns:
+        None: Nothing, changes are performed in-place.
+    """
+    if "SPDXID" not in sbom:
+        sbom["SPDXID"] = "SPDXRef-DOCUMENT"
+    if "dataLicense" not in sbom:
+        sbom["dataLicense"] = "CC0-1.0"
+    if "spdxVersion" not in sbom:
+        sbom["spdxVersion"] = "SPDX-2.3"
+    if "documentNamespace" not in sbom:
+        sbom["documentNamespace"] = f"https://konflux-ci.dev/spdx/{uuid.uuid4()}"
+    if "name" not in sbom:
+        sbom["name"] = "MOBSTER:UNFILLED_NAME (please update this field)"
+
+    creation_info = sbom.get("creationInfo", {})
+    if "created" not in creation_info:
+        creation_info["created"] = "1970-01-01T00:00:00Z"
+    creators = creation_info.get("creators", [])
+    new_creators = [await normalize_actor(creator) for creator in creators]
+    if not new_creators:
+        new_creators.append(f"Tool: Mobster-{get_mobster_version()}")
+    creation_info["creators"] = new_creators
+    sbom["creationInfo"] = creation_info
+
+    for package in sbom.get("packages", []):
+        await normalize_package(package)
+
+
+async def find_spdx_root_relationships(sbom: Document) -> list[Relationship]:
+    """
+    Finds the relationship describing the root element.
+    Args:
+        sbom (spdx_tools.spdx.model.document.Document): The SBOM
+
+    Returns:
+        spdx_tools.spdx.model.relationship.Relationship: The root relationship
+    """
+    relationships = []
+    for relationship in sbom.relationships:
+        for relationship_type in (
+            RelationshipType.DESCRIBES,
+            RelationshipType.DESCRIBED_BY,
+        ):
+            # The root element is either DESCRIBED_BY SPDXRef-DOCUMENT
+            # or SPDXRef-DOCUMENT DESCRIBES the root element
+            if relationship.relationship_type is relationship_type:
+                relationships.append(relationship)
+    return relationships
+
+
+async def find_spdx_root_packages_spdxid(sbom_doc: Document) -> list[str]:
+    """
+    Finds the root element of an SPDX SBOM and returns its SPDXID.
+    Args:
+        sbom_doc (spdx_tools.spdx.model.document.Document): The SBOM
+
+    Returns:
+        str: The SPDXID of the root package
+    """
+    spdx_ids = []
+    root_relationships = await find_spdx_root_relationships(sbom_doc)
+    for root_relationship in root_relationships:
+        if root_relationship.relationship_type is RelationshipType.DESCRIBES:
+            spdx_ids.append(root_relationship.related_spdx_element_id)
+        else:
+            spdx_ids.append(root_relationship.spdx_element_id)
+    return spdx_ids  # type: ignore[return-value]
```

Comment:
We should not really ignore types here and everywhere else in this PR. This function returns `list[Any]` but its type doesn't says that.

Reply:
I have fixed this line in the code by filtering out non-`str` relationship references (those are not relevant for us anyway). In some places though we know the structure returned will adhere to the SPDX/CycloneDX specification but mypy cannot assess that. So I think it is unavoidable to use this here and there if we want to provide useful type hints.